api/analysis.go

```diff
@@ -0,0 +1,927 @@
+		q = q.Where("ID IN (?)", tags)
+	}
+	return
+}
+
+//
+// analysisIDs provides analysis IDs.
+func (h *AnalysisHandler) analysisIDs(ctx *gin.Context, f *qf.Filter) (q *gorm.DB) {
+	q = h.DB(ctx)
+	q = q.Model(&model.Analysis{})
+	q = q.Select("MAX(ID)")
+	q = q.Where("ApplicationID IN (?)", h.appIDs(ctx, f))
+	q = q.Group("ApplicationID")
+	return
+}
+
+//
+// rulesetIDs provides ruleSet IDs.
+func (h *AnalysisHandler) rulesetIDs(ctx *gin.Context, f *qf.Filter) (q *gorm.DB) {
+	q = h.DB(ctx)
+	q = q.Model(&model.AnalysisRuleSet{})
+	q = q.Select("ID")
+	q = q.Where("AnalysisID IN (?)", h.analysisIDs(ctx, f))
+	techFilter := f.Resource("technology")
+	if field, found := techFilter.Field("source"); found {
+		field = field.As("Name")
+		tech := h.DB(ctx)
+		tech = tech.Model(&model.AnalysisTechnology{})
+		tech = tech.Select("RuleSetID")
+		tech = tech.Where("Source", true)
+		tech = tech.Where(field.SQL())
+		q = q.Where("ID IN (?)", tech)
+	}
+	if field, found := techFilter.Field("target"); found {
+		field = field.As("Name")
+		tech := h.DB(ctx)
+		tech = tech.Model(&model.AnalysisTechnology{})
+		tech = tech.Select("RuleSetID")
+		tech = tech.Where("Source", false)
+		tech = tech.Where(field.SQL())
+		q = q.Where("ID IN (?)", tech)
+	}
+	return
+}
+
+//
+// depAppMap provides map of deps to application refs.
+func (h *AnalysisHandler) depAppMap(ctx *gin.Context, f *qf.Filter) (mp map[string][]Ref, err error) {
+	type AppRef struct {
+		model.AnalysisDependency
+		AppID   uint
+		AppName string
+	}
+	mp = make(map[string][]Ref)
+	db := h.DB(ctx)
+	db = db.Raw(
+		strings.Join([]string{
+			"SELECT DISTINCT ",
+			"  app.ID AppID,",
+			"  app.Name AppName,",
+			"  dep.Name,",
+			"  dep.Version,",
+			"  dep.SHA",
+			"FROM",
+			"  AnalysisDependency dep,",
+			"  Analysis an,",
+			"  Application app",
+			"WHERE ",
+			"  app.ID = an.ApplicationID",
+			"  AND an.ID = dep.AnalysisID",
+			"  AND an.ID IN (?)",
+		},
+			" "),
+		h.analysisIDs(ctx, f))
+	list := make([]AppRef, 0, 0)
+	result := db.Find(&list)
+	if result.Error != nil {
+		err = result.Error
+		return
+	}
+	for i := range list {
+		m := &list[i]
+		mp[m.Key()] = append(
+			mp[m.Key()],
+			Ref{ID: m.AppID, Name: m.AppName})
+	}
+	return
+}
+
+//
+// issueAppMap provides map of issues to application refs.
+func (h *AnalysisHandler) issueAppMap(ctx *gin.Context, f *qf.Filter) (mp map[string][]Ref, err error) {
+	type AppRef struct {
+		AppID   uint
+		AppName string
+		RuleID  string
+	}
+	mp = make(map[string][]Ref)
+	db := h.DB(ctx)
+	db = db.Raw(
+		strings.Join([]string{
+			"SELECT DISTINCT",
+			"  app.ID AppID,",
+			"  app.Name AppName,",
+			"  issue.RuleID",
+			"FROM",
+			"  AnalysisIssue issue,",
+			"  AnalysisRuleSet ruleset,",
+			"  Analysis an,",
+			"  Application app ",
+			"WHERE",
+			"  app.ID = an.ApplicationID",
+			"  AND an.ID = ruleset.AnalysisID",
+			"  AND ruleset.ID = issue.RuleSetID",
+			"  AND an.ID IN (?)",
+		},
```

Comment:
**hint**: This WHERE could likely be narrowed a little more by applying the issue filter but decided not to for the first iteration out of simplicity.  Not sure it's worth it.